src/coreclr/src/vm/jithelpers.cpp

```diff
@@ -3001,84 +3003,66 @@ HCIMPL3(VOID, JIT_Unbox_Nullable, void * destPtr, CORINFO_CLASS_HANDLE type, Obj
 HCIMPLEND
 
 /*************************************************************/
-/* framed helper that handles full-blown type equivalence */
-NOINLINE HCIMPL2(LPVOID, JIT_Unbox_Helper_Framed, CORINFO_CLASS_HANDLE type, Object* obj)
+/* framed Unbox helper that handles enums and full-blown type equivalence */
+NOINLINE HCIMPL2(LPVOID, Unbox_Helper_Framed, MethodTable* pMT1, Object* obj)
 {
     FCALL_CONTRACT;
 
     LPVOID result = NULL;
+    MethodTable* pMT2 = obj->GetMethodTable();
 
     OBJECTREF objRef = ObjectToOBJECTREF(obj);
     HELPER_METHOD_FRAME_BEGIN_RET_1(objRef);
-    if (TypeHandle(type).IsEquivalentTo(objRef->GetTypeHandle()))
+
+    CorElementType corElementType = pMT1->GetInternalCorElementType();
+    if (corElementType != ELEMENT_TYPE_VALUETYPE &&  corElementType == pMT2->GetInternalCorElementType())
+    {
+        // we allow enums and their primitive type to be interchangable
+        result = objRef->GetData();
+    }
+    else if (pMT1->IsEquivalentTo(pMT2))
     {
         // the structures are equivalent
         result = objRef->GetData();
     }
     else
     {
-        COMPlusThrowInvalidCastException(&objRef, TypeHandle(type));
+        COMPlusThrowInvalidCastException(&objRef, TypeHandle(pMT1));
     }
+
+    HELPER_METHOD_POLL(); 
```

Comment:
Result is not protected here. Move the POLL right after `HELPER_METHOD_FRAME_BEGIN_RET_1`

Reply:
Or protect the result, but moving the poll earlier is simpler.

---
In reply to: [379897330](https://github.com/dotnet/runtime/pull/32353#discussion_r379897330) [](ancestors = 379897330)